src/LinearA.hs

```diff
@@ -518,73 +518,105 @@ shuffle scope es = go [] [] (freshVars scope) es
 -------------------- Unzip --------------------
 
 unzipProgram :: Program -> Program
-unzipProgram (Program funcMap) = Program $ flip M.foldMapWithKey funcMap $ \name def -> do
-  let (udef, udef') = unzipFunc def
-  M.fromList [(name ++ ".nonlin", udef), (name ++ ".lin", udef')]
-
-unzipFunc :: FuncDef -> (FuncDef, FuncDef)
-unzipFunc (FuncDef formalsWithTys linFormalsWithTys (MixedType retTys linRetTys) body) =
+unzipProgram orig@(Program funcMap) = new where
+  new = Program $ flip M.foldMapWithKey funcMap $ \name def -> do
+    -- Tying the knot by passing `new` here does assume that laziness
+    -- will figure out a sensible order in which to actually perform
+    -- the unzipping so that the type of every unzipped callee is
+    -- available for its unzipped caller.
```

Comment:
Which means that this will succeed for as long as there is no (mutual) recursion in the program. On the other hand, given that we don't have conditionals in the language, any such program would necessarily fail to terminate anyway, but it is a bit funny that our compiler would fail to terminate as well. In any case, I think this is ok for now.